File: src/main/kotlin/카카오코테_다트게임.java
import java.io.BufferedWriter;
import java.io.OutputStreamWriter;
import java.util.ArrayList;

public class 카카오코테_다트게임 {
    public static void main(String[] args) {
        ArrayList<Integer> arrayNum = new ArrayList<>();
        try {
            int answer = 0;

            String dartResult = "1D2S#10S";

            int i = 0;
            int cnt = 0;
            while (i < dartResult.length()) {
                if (Character.isDigit(dartResult.charAt(i))) {
                    cnt++;
                    if (Character.isDigit(dartResult.charAt(i+1))) {
                        arrayNum.add(Integer.parseInt("10"));
                        i+=2;
                        continue;
                    }
                    arrayNum.add(Integer.parseInt(String.valueOf(dartResult.charAt(i))));
                    i++;
                    continue;
                }

                if (dartResult.charAt(i) == 'S') {
                    i++;
                    continue;
                }
                if (dartResult.charAt(i) == 'D') {
                    int a = (int) Math.pow(arrayNum.get(cnt-1), 2);
                    arrayNum.set(cnt-1, a);
                    i++;
                    continue;
                }
                if (dartResult.charAt(i) == 'T') {
                    int a = (int) Math.pow(arrayNum.get(cnt-1), 3);
                    arrayNum.set(cnt-1, a);
                    i++;
                    continue;
                }
                if (dartResult.charAt(i) == '*') {
                    i++;
                    if (cnt > 1) {
                        int a = arrayNum.get(cnt-2) * 2;
                        arrayNum.set(cnt-2, a);
                    }
                    int a = arrayNum.get(cnt-1) * 2;
                    arrayNum.set(cnt-1, a);
                    continue;
                }
                if (dartResult.charAt(i) == '#') {
                    i++;
                    int a = arrayNum.get(cnt-1) * -1;
                    arrayNum.set(cnt-1, a);
                }
            }

            BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
            bw.write(arrayNum.get(0) + arrayNum.get(1) + arrayNum.get(2) +"");
            bw.flush();
            bw.close();

        } catch (Exception e) {
            e.printStackTrace();
        } finally {
//            return arrayNum.get(0) + arrayNum.get(1) + arrayNum.get(2);
        }

    }
}

File: src/main/kotlin/카카오코테_다트게임.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*
import kotlin.math.pow
import kotlin.math.sqrt

fun main(args: Array<String>) {

    try {
        val br = BufferedReader(InputStreamReader(System.`in`))
        val bw = BufferedWriter(OutputStreamWriter(System.out))

        var input = br.readLine()

        var arrayNum = mutableListOf<Int>()

        var i = 0
        var cnt = 0
        run {
            while (i < input.length) {
                // 문자(char)를 toInt()하면 아스키코드로 전환, 문자열(string)을 toInt()하면 숫자로 변환
                if (input[i].toInt() in 48..57) { // 문자가 숫자일경우
                    cnt++
                    if (input[i+1].toInt() in 48 .. 57) {   // 2자리 숫자일 경우
                        var num = ("${input[i]}" + "${input[i+1]}").toInt()
                        arrayNum.add(num)
                        i+=2    // 2자리 숫자라 다음 숫자까지 입력이 되었기때문에 다음숫자 index는 뛰어 넘기기위해 +2를 해준다
                        continue
                    }
                    arrayNum.add(input[i].toString().toInt())
                    i++
                    continue
                }

                if (input[i].equals('S')) {
                    arrayNum[cnt-1] = arrayNum[cnt-1].toFloat().pow(1).toInt()
                    i++
                    continue
                }

                if (input[i].equals('D')) {
                    arrayNum[cnt-1] = arrayNum[cnt-1].toFloat().pow(2).toInt()
                    i++
                    continue
                }

                if (input[i].equals('T')) {
                    arrayNum[cnt-1] = arrayNum[cnt-1].toFloat().pow(3).toInt()
                    i++
                    continue
                }

                if (input[i].equals('*')) {
                    i++
                    if (cnt>1) {
                        arrayNum[cnt-2] *= 2
                    }
                    arrayNum[cnt-1] *= 2
                    continue
                }

                if (input[i].equals('#')) {
                    i++
                    arrayNum[cnt-1] *= -1
                }
            }
        }

        bw.write("${arrayNum[0]+arrayNum[1]+arrayNum[2]}")
        bw.flush()
        bw.close()



    } catch (e: Exception) {
        e.printStackTrace()
    }

}